sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -0,0 +1,334 @@
+                processBundleInstructionId,
+                target,
+                coderSpec,
+                beamFnDataClient);
+        addStartFunction.accept(remoteGrpcWriteRunner::registerForOutput);
+        consumer = remoteGrpcWriteRunner::consume;
+        addFinishFunction.accept(remoteGrpcWriteRunner::close);
+        break;
+
+      case DATA_INPUT_URN:
+        target = BeamFnApi.Target.newBuilder()
+            .setPrimitiveTransformReference(primitiveTransform.getId())
+            .setName(getOnlyElement(primitiveTransform.getInputsMap().keySet()))
+            .build();
+        coderSpec = (BeamFnApi.Coder) fnApiRegistry.apply(
+            getOnlyElement(primitiveTransform.getOutputsMap().values()).getCoderReference());
+        BeamFnDataReadRunner<OutputT> remoteGrpcReadRunner =
+            new BeamFnDataReadRunner<>(
+                functionSpec,
+                processBundleInstructionId,
+                target,
+                coderSpec,
+                beamFnDataClient,
+                outputMap);
+        addStartFunction.accept(remoteGrpcReadRunner::registerInputLocation);
+        consumer = null;
+        addFinishFunction.accept(remoteGrpcReadRunner::blockTillReadFinishes);
+        break;
+
+      case JAVA_DO_FN_URN:
+        DoFnRunner<InputT, OutputT> doFnRunner = createDoFnRunner(functionSpec, outputMap);
+        addStartFunction.accept(doFnRunner::startBundle);
+        addFinishFunction.accept(doFnRunner::finishBundle);
+        consumer = doFnRunner::processElement;
+        break;
+
+      case JAVA_SOURCE_URN:
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        BoundedSourceRunner<BoundedSource<OutputT>, OutputT> sourceRunner =
+            createBoundedSourceRunner(functionSpec, outputMap);
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        ThrowingConsumer<WindowedValue<?>> sourceConsumer =
+            (ThrowingConsumer)
+                (ThrowingConsumer<WindowedValue<BoundedSource<OutputT>>>)
+                    sourceRunner::runReadLoop;
+        // TODO: Remove and replace with source being sent across gRPC port
+        addStartFunction.accept(sourceRunner::start);
+        consumer = (ThrowingConsumer) sourceConsumer;
+        break;
+    }
+
+    if (consumer != null) {
+      for (Map.Entry<String, BeamFnApi.Target.List> entry :
+          primitiveTransform.getInputsMap().entrySet()) {
+        for (BeamFnApi.Target target : entry.getValue().getTargetList()) {
+          addConsumer.accept(target, consumer);
+        }
+      }
+    }
+  }
+
+  public BeamFnApi.InstructionResponse.Builder processBundle(BeamFnApi.InstructionRequest request)
+      throws Exception {
+    BeamFnApi.InstructionResponse.Builder response =
+        BeamFnApi.InstructionResponse.newBuilder()
+            .setProcessBundle(BeamFnApi.ProcessBundleResponse.getDefaultInstance());
+
+    long bundleId = request.getProcessBundle().getProcessBundleDescriptorReference();
```

Comment:
what, if any, is the danger of bundleId collisions? Assuming they're generated randomly, I can easily imagine collisions if we use `long` -- because we'll run more than 2^32 bundles in big jobs.